tests/net/tcp2/src/main.c

```diff
@@ -1292,6 +1324,112 @@ static void test_client_invalid_rst(void)
 	check_rst_succeed(ctx, wnd - 1);
 	check_rst_succeed(NULL, 0);
 	check_rst_succeed(NULL, 1);
+
+	net_tcp_put(ctx);
+}
+
+#define MAX_DATA 100
+static uint32_t expected_ack = MAX_DATA + 1;
+static struct net_context *ooo_ctx;
+
+static void handle_server_recv_out_of_order(struct net_pkt *pkt)
+{
+	struct tcphdr th;
+	int ret;
+
+	ret = read_tcp_header(pkt, &th);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	/* Verify that we received all the queued data */
+	zassert_equal(expected_ack, ntohl(th.th_ack),
+		      "Not all pending data received. "
+		      "Expected ACK %u but got %u",
+		      expected_ack, ntohl(th.th_ack));
+
+	test_sem_give();
+
+	return;
+
+fail:
+	zassert_true(false, "%s failed", __func__);
+	net_pkt_unref(pkt);
+}
+
+static void test_server_recv_out_of_order_data(void)
```

Comment:
This test just verifies that packets can be received in reversed order (5, 4, 3, 2, 1, 0). What if we received packets in this order: 3, 1, 0, 5, 4, 2?

Reply:
> What if we received packets in this order: 3, 1, 0, 5, 4, 2?

That complicates things a lot so that I did not implement it in this first version. Now the new packet received can be either right before an earlier packet or a later one, so no holes are allowed in the seq list.

Reply:
I think we need to describe current behavior (limitation?) in both Kconfig and commit message. Otherwise people will expect frames received like in example above (3, 1, 0, 5, 4, 2) to work correctly, as they are out of order. Currently only a subset of "out of order" patterns are handled.

>That complicates things a lot so that I did not implement it in this first version. Now the new packet received can be either right before an earlier packet or a later one, so no holes are allowed in the seq list.

I was thinking a bit about that "hole handling" and/or making that TCP queueing more generic. My idea would be to queue `net_pkt` instead of `net_buf`. This could be accomplished by placing `struct rbnode rbnode;` in the union inside `struct net_pkt` next to `struct k_work work;` and `intptr_t sock_recv_fifo;`. Then putting all packets to rbtree, so the packet with lowest sequence number could be retrieved with `rb_get_min()` with pretty low cost. This would have following advantages over current design:
- no need to traverse each `net_buf` and calculate `seq` for them,
- no need to keep `seq` in `net_buf` user data,
- no need to have an overhead of one dedicated `struct net_pkt` per each TCP connection (which keeps track of out-of-order buffers),
- handling of "holes",
- cleaner implementation (I believe) as:
  - rbtree would take care of sorting packets in a proper order, 
  - seq does not need to be calculated for each net_buf - it is enough to check of first (with minimum seq) rbnode has the seq which should be received next).

There would be also a downside:
- `net_pkt` could not be freed, when putting packets on queue; instead, then would be freed only after receiving all previous packets (or if timeout expires).

Reply:
> My idea would be to queue `net_pkt` instead of `net_buf`.

The TCP data is in net_buf so we need to queue those too.

I am all +1 if you have a good plan, what you describe sounds usable although quite a memory hog if we would need to queue the net_pkt too.

Reply:
>The TCP data is in net_buf so we need to queue those too.

Of course. I just meant to queue and sort `net_pkt` objects, so all `net_buf` would be still attached to `net_pkt` and sorted "implicitly".

Reply:
> I just meant to queue and sort `net_pkt` objects, so all `net_buf` would be still attached to `net_pkt` and sorted "implicitly".

Ok, understood. Well, that is certainly possible, I am just worried about the memory consumption if we need to keep all the net_pkt pending for some time. Patches are welcome if you want to experiment with this.